File: frontend/src/api/recognition.ts
import { apiGet, apiPost } from "./client";

export type RecognitionRunResult = {
  sheet_id: number;
  status: string;
  run_type: string;
  output_path: string | null;
  text_length: number;
  error_code: string | null;
  error_message: string | null;
};

export type BatchRecognitionResult = {
  batch_id: number;
  total_count: number;
  success_count: number;
  failed_count: number;
  items: RecognitionRunResult[];
};

export type RecognitionRun = {
  id: number;
  project_id: number;
  batch_id: number;
  file_id: number | null;
  sheet_id: number;
  run_type: string;
  engine_name: string;
  engine_version: string;
  status: string;
  output_path: string | null;
  started_at: string;
  finished_at: string;
  error_code: string | null;
  error_message: string | null;
  created_at: string;
};

const LONG_OP_TIMEOUT_MS = 300_000;

export function extractSheetText(sheetId: number): Promise<RecognitionRunResult> {
  return apiPost<RecognitionRunResult>(`/api/sheets/${sheetId}/extract-text`, undefined, {
    timeoutMs: LONG_OP_TIMEOUT_MS
  });
}

export function ocrSheetTitle(sheetId: number): Promise<RecognitionRunResult> {
  return apiPost<RecognitionRunResult>(`/api/sheets/${sheetId}/ocr-title`, undefined, {
    timeoutMs: LONG_OP_TIMEOUT_MS
  });
}

export function extractBatchText(batchId: number): Promise<BatchRecognitionResult> {
  return apiPost<BatchRecognitionResult>(`/api/imports/${batchId}/extract-text`, undefined, {
    timeoutMs: LONG_OP_TIMEOUT_MS
  });
}

export type OcrJobStatus = {
  batch_id: number;
  status: "idle" | "running" | "completed" | "failed";
  total: number;
  processed: number;
  success_count: number;
  failed_count: number;
  started_at: string | null;
  finished_at: string | null;
  message: string | null;
};

export function ocrBatchTitles(batchId: number): Promise<OcrJobStatus> {
  return apiPost<OcrJobStatus>(`/api/imports/${batchId}/ocr-titles`, undefined);
}

export function getOcrBatchJob(batchId: number): Promise<OcrJobStatus> {
  return apiGet<OcrJobStatus>(`/api/imports/${batchId}/ocr-job`);
}

export function listRecognitionRuns(sheetId: number): Promise<RecognitionRun[]> {
  return apiGet<RecognitionRun[]>(`/api/sheets/${sheetId}/recognition-runs`);
}

File: backend/api/recognition.py
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from backend.core.database import get_db
from backend.schemas.recognition_run import (
    BatchRecognitionResult,
    OcrJobStatus,
    RecognitionRunRead,
    RecognitionRunResult,
)
from backend.services import ocr_service, pdf_text_service, recognition_run_service

router = APIRouter(prefix="/api", tags=["recognition"])


@router.post("/sheets/{sheet_id}/extract-text", response_model=RecognitionRunResult)
def extract_sheet_text(sheet_id: int, db: Session = Depends(get_db)) -> RecognitionRunResult:
    return pdf_text_service.extract_text_for_sheet(db, sheet_id)


@router.post("/sheets/{sheet_id}/ocr-title", response_model=RecognitionRunResult)
def ocr_sheet_title(sheet_id: int, db: Session = Depends(get_db)) -> RecognitionRunResult:
    return ocr_service.ocr_title_for_sheet(db, sheet_id)


@router.post("/imports/{batch_id}/extract-text", response_model=BatchRecognitionResult)
def extract_batch_text(batch_id: int, db: Session = Depends(get_db)) -> BatchRecognitionResult:
    return pdf_text_service.extract_text_for_batch(db, batch_id)


@router.post("/imports/{batch_id}/ocr-titles", response_model=OcrJobStatus)
def ocr_batch_titles(batch_id: int, db: Session = Depends(get_db)) -> OcrJobStatus:
    return ocr_service.start_ocr_batch_job(db, batch_id)


@router.get("/imports/{batch_id}/ocr-job", response_model=OcrJobStatus)
def get_ocr_batch_job(batch_id: int, db: Session = Depends(get_db)) -> OcrJobStatus:
    return ocr_service.get_ocr_batch_job(db, batch_id)


@router.get("/sheets/{sheet_id}/recognition-runs", response_model=list[RecognitionRunRead])
def list_sheet_recognition_runs(
    sheet_id: int,
    db: Session = Depends(get_db),
) -> list[RecognitionRunRead]:
    return recognition_run_service.list_sheet_runs(db, sheet_id)
